### DISCUSSION TITLE
Difficulties with void plugins

### DISCUSSION THREAD

**Discussion by johncin** [react-page/react-page#1243] [Q&A] [answered]
Here's a repo with what I've been trying to get to work. https://github.com/johncin/reactpage-admin-test

Basically, I want to use reactpage to create pages that can consume json data from a rest api. So, I've created a layout component where you set an apiURL and then it supplies the resulting data to its children through a react context. So far so good. My issues come with the "apiField" cell plugin that takes in the name of the data property to display and then display said data. Just for this example it displays the data from the first object in the json data array. It's a void plugin of type "inline". I can add this field once and it displays correctly. But if I add another, it throws errors about "Cannot resolve a DOM node from Slate node". And I've googled this a lot, trying many things to get this to work, but it just won't work no matter what I try... I'm really at my wit's end.

I should also note that I couldn't get one of the reactpage components for react-admin to work. RPageInput. I had to copy it and change one of the variables to get it to work. In particular this part
`const {
  field: { value, onChange },
} = useInput({ source })`

**Comment by macrozone** [comment]
hi @johncin I briefly looked at your example.

It looks like you want to do add page-links to slate.

In your case, you use a `ApiFieldSelector` which is in my opinion too generic. Say, you want to have links to a `Product`-page, so you create a `ProductIdSelectField`. This field is an input, where the value is a productId (string). onChange is called with the selected productId. How this field fetches all products to select is this component's responsibility, not ReactPage. Typically, you would use `react-query`, `apollo-client` or `swr` etc. (or just use "fetch-in-useEffect", altough personally i recomend to always use proper data-fetching libraries that handle caching and batching properly). 
I would avoid having a specific react-page cell component with the provider that loads specific data. If you need a provider ( like with ApolloProvider), wrap your whole app with it, or the editor.

Next step is to create a slate-plugin, which you already started. You used type "inline", I used `mark` in my projects. Also this example here uses `mark`: https://github.com/react-page/react-page/blob/master/examples/pages/examples/reactadmin.tsx#L103 
tbh. I don't know why its better to use `mark` in this case, but its better for links.

Next step is to have a `ProductLink`-component, which receives the productId (given the example with products), and renders a link for it, e.g. look at this example: https://github.com/react-page/react-page/blob/master/examples/pages/examples/reactadmin.tsx#L131

Now to the dreaded "Cannot resolve a DOM node from Slate node"-error. This is probably because you render aditional things after the text that the user providers:

```
 <TypeToUse {...{ ...attributes, ...linkAttributes }}>
        {children}
        <span contentEditable={false} style={{ position: 'relative' }}>
          <span style={{ userSelect: 'none' }}>{fieldText}</span>
        </span>
      </TypeToUse>
```


ideally, you only render user provided text (just render children). I also thought that setting `contentEditable={false}` to anything after `children` should prevent the error above from happening 🤔 

If you DON't want to render children and only render `fieldText`, then yes, you should be able to set `isVoid: true` and do not render the children at all.

**Reply by macrozone** [reply]
@johncin sorry for the long brain dump. I now try to reproduce the problem with the "Cannot resolve a DOM node from Slate node" for this specific case

**Comment by macrozone** [comment, marked as answer]
@johncin i can reproduce the issue and found a fix! Will release it soon.

**Reply by macrozone** [reply]
@johncin can you try version 5.1.7 ?

**Reply by johncin** [reply]
It works now! Oh wow that's amazing. Thanks a bunch!

At first I was getting an error message in my CellPlugins array, but after deleting node_modules and yarn.lock and installing again it was all fine.

**Reply by macrozone** [reply]
on a side note: you don't have to render `children` in your void-plugin's component (this is handled outside), in fact, that is just undefined. same with `attributes`
